### PR TITLE
py/modsys: Initial implementation of sys.getsizeof().

### PR DESCRIPTION
**Pull request by pfalcon**

### CODE REVIEW

**Comment by dpgeorge** [comment]
I guess the usual comments apply: it adds code size and so far we haven't seen any need/use for it.  Also, to be useful it would need to support more built-in types, like dict, tuple, str, bytes, bytearray, which is only going to add more code.  And it still won't be useful for native types that are not built in, some of which are the real culprits when it comes to heap usage (eg ssl socket instances).

That said, if it's enabled only on unix-coverage then it's a nice function to have.

Note: it looks like minimal build doesn't use gc-sections so that could be why it increases in size for this patch.

**Comment by pfalcon** [comment]
> it adds code size

That's why it's not enabled by default even for unix port (only for coverage build).

> we haven't seen any need/use for it

Well, the need for memory profiling/diagnosing is there right from the start, it's just we don't do much for it, and users manage to not see glaring need for it either. As I mentioned, I'm rather surprised to see request like https://github.com/micropython/micropython/issues/3009 (want to debug? debug with CPython) instead of requests to have more insight into memory allocation/management. We have https://github.com/micropython/micropython/issues/996 open for a long time and this patch can be seen as step in that direction.

**Comment by pfalcon** [comment]
> And it still won't be useful for native types that are not built in, some of which are the real culprits when it comes to heap usage (eg ssl socket instances).

So, this PR is to show/discuss approach to implementing it. The current patch already shows 2 approaches:

1. Keep the code for a particular object at sys.getsizeof() code.
2. Keep the code at object's codebase, and call it explicitly.

This can be extends further:
3. Virtualize in efficient manner, e.g. add a unary_op for that.
4. Virtualize in Python-level extensible manner, e.g. implement `__sizeof__` 
method https://docs.python.org/3/library/sys.html#sys.getsizeof

The question is which of these to use. IMHO, single method won't work, a combination should be used. But if we choose to use 3, then maybe 2 shouldn't be used. And 4, as having the most overhead, should be put off until really needed.

**Comment by dpgeorge** [comment]
> 3. Virtualize in efficient manner, e.g. add a unary_op for that.

This might be the cleanest and most code-compact way to do it.  Having an additional MP_UNARY_OP means that there won't (in some cases, not for ussl...) need to be any additional C functions, which leads to the most code overhead.  Some built-in classes (eg list, dict) could be done inline in the sys_getsizeof to save space, but it may turn out to be more compact to put them in their own unary_op method because then you don't need to do any type checking in sys_getsizeof, it would directly go to type->unary_op.

It would be interesting to see if the MP_UNARY_OP way would give smaller (or similar) code size to the patch proposed here.

**Comment by pfalcon** [comment]
> It would be interesting to see if the MP_UNARY_OP way would give smaller (or similar) code size to the patch proposed here.

Indeed, it does:

1st adhoc verison: 393268

Using MP_UNARY_OP: 393244

Will push new version soon.

**Comment by pfalcon** [comment]
This now supports instances, list and dict, which should be a good start. So, I'd like to merge this.

File: py/objdict.c [hunk 1]
```diff
@@ -105,6 +105,12 @@ STATIC mp_obj_t dict_unary_op(mp_uint_t op, mp_obj_t self_in) {
     switch (op) {
         case MP_UNARY_OP_BOOL: return mp_obj_new_bool(self->map.used != 0);
         case MP_UNARY_OP_LEN: return MP_OBJ_NEW_SMALL_INT(self->map.used);
+        #if MICROPY_PY_SYS_GETSIZEOF
+        case MP_UNARY_OP_SIZEOF: {
+            mp_int_t sz = sizeof(*self) + sizeof(*self->map.table) * self->map.alloc;
```
Comment (dpgeorge) [inline py/objdict.c]: It should probably be size_t type, since it counts bytes.

File: py/objtype.c [hunk 2]
```diff
@@ -87,6 +87,16 @@ STATIC int instance_count_native_bases(const mp_obj_type_t *type, const mp_obj_t
     }
 }
 
+mp_uint_t mp_obj_instance_sizeof(mp_obj_t obj) {
```
Comment (dpgeorge) [inline py/objtype.c]: Why not make this static?

Comment (dpgeorge) [inline py/objtype.c]: Should return size_t type.

File: unix/mpconfigport.h [hunk 3]
```diff
@@ -96,6 +96,7 @@
 #define MICROPY_PY_SYS_MAXSIZE      (1)
 #define MICROPY_PY_SYS_STDFILES     (1)
 #define MICROPY_PY_SYS_EXC_INFO     (1)
+#define MICROPY_PY_SYS_GETSIZEOF    (0)
```
Comment (dpgeorge) [inline unix/mpconfigport.h]: For config options that are left at their default value (in this case disabled) I prefer to leave them out of the config file.

File: py/objdict.c (hunk 1 above)
Comment (pfalcon) [inline py/objdict.c]: Fixed. This is just another illustration of concerns expressed in https://github.com/micropython/micropython/pull/3111 - some people didn't master even size_t usage, and there're ideas to introduce even more obscure stuff ;-).

File: py/objtype.c (hunk 2 above)
Comment (pfalcon) [inline py/objtype.c]: Now can even be inlined at the call site.

Fixed.

File: unix/mpconfigport.h (hunk 3 above)
Comment (pfalcon) [inline unix/mpconfigport.h]: This is kind of invite (to individual users) to enable it, but ok.

**Comment by pfalcon** [comment]
Merged.

**Pull request closed without merging**